modules/backup.nix

```diff
@@ -0,0 +1,190 @@
+{ lib
+, config
+, pkgs
+, ...
+}:
+
+let
+  cfg = config.services.backup;
+
+  mkZfsPreHook = mountpoint: ''
+    DATASET="$(findmnt -nr -o -target -T "${mountpoint}")"
+    zfs snapshot -r "$DATASET@borg"
+
+    # https://github.com/borgbackup/borg/issues/6652
+    ls $MOUNTPOINT/.zfs/snapshot/backup/ > /dev/null
+  '';
+
+  mkZfsPostHook = mountpoint: ''
+    DATASET="$(findmnt -nr -o -target -T "${mountpoint}")"
+    zfs destroy -r "$DATASET@borg"
+  '';
+in
+{
+  options.services.backup = with lib; with types; {
+    user = mkOption {
+      type = str;
+      description = ''
+        Username for the SSH remote host.
+      '';
+    };
+
+    host = mkOption {
+      type = str;
+      description = ''
+        Hostname of the SSH remote host.
+      '';
+    };
+
+    hostPublicKey = mkOption {
+      type = str;
+      example = "ssh-rsa AAAAB3NzaC1yc2EAAAABIwAAAQEA5EB5p/5Hp3hGW1oHok+PIOH9Pbn7cnUiGmUEBrCVjnAw+HrKyN8bYVV0dIGllswYXwkG/+bgiBlE6IVIBAq+JwVWu1Sss3KarHY3OvFJUXZoZyRRg/Gc/+LRCE7lyKpwWQ70dbelGRyyJFH36eNv6ySXoUYtGkwlU5IVaHPApOxe4LHPZa/qhSRbPo2hwoh0orCtgejRebNtW5nlx00DNFgsvn8Svz2cIYLxsPVzKgUxs8Zxsxgn+Q/UvR7uq4AbAhyBMLxv7DjJ1pc7PJocuTno2Rw9uMZi1gkjbnmiOh6TTXIEWbnroyIhwc8555uto9melEUmWNQ+C+PwAK+MPw==";
+      description = ''
+        Public SSH host key of the remote host. Discoverable using e.g. `ssh-keyscan`.
+      '';
+    };
+
+    port = mkOption {
+      type = port;
+      default = 22;
+      description = ''
+        Port of the SSH remote host.
+      '';
+      apply = toString;
+    };
+
+    sshKey = mkOption {
+      type = path;
+      example = "/var/keys/ssh-key";
+      description = ''
+        Path to the SSH key required to access the remote host.
+      '';
+    };
+
+    secretPath = mkOption {
+      type = path;
+      example = "/var/keys/borg-secret";
+      description = ''
+        Path to the secret used to encrypt backups in the repository.
+      '';
+    };
+
+    quota = mkOption {
+      type = nullOr str;
+      default = null;
+      example = "90G";
+      description = ''
+        Quota for the borg repository. Useful to prevent the target disk from running full and ensuring borg keeps some space to work with.
+      '';
+    };
+
+    includes = mkOption {
+      type = listOf path;
+      default = [];
+      description = ''
+        Paths to include in the backup.
+      '';
+    };
+    includesZfsDatasets = mkOption {
+      type = listOf str;
+      default = [];
+      description = ''
+        ZFS datasets referenced by mountpoint to snapshot and include
+      '';
+    };
```

Comment:
Not too happy about the name. Taking the mountpoint is necessary to declaratively include it in the backup paths. I'm resolving the device name using `findmnt`.